### PR TITLE
Add functionality to utilise the .forceignore file for exclusions  

### PR DESCRIPTION
**Pull request by jjulicher**
## What does this pull request contains? Explain your changes.
Add functionality to utilise the .forceignore file for exclusions  as lwc directory can contain files that are not to be included in the package.xml. Includes additional CLI parameter -i / --ignore to specify file .
---

- [x] Added for new features.
- [ ] Changed for changes in existing functionality.
- [ ] Deprecated for soon-to-be removed features.
- [ ] Removed for now removed features.
- [ ] Fixed for any bug fixes.
- [ ] Security in case of vulnerabilities.

## Explain your changes

---

Contains additional config & tests to filter using ignore functionality for the repoGitDiff util. 
Added in command line parameters to specify the ignore file. 

## Does this close any currently open issues?

---

no

- [ ] Jest test to check the fix is applied are added.

## Any particular element to being able to test locally

---

_(Write your answer here.)_

## Any other comments?

---

_(Write your answer here.)_

## Where has this been tested?

---

**Operating System:** …
Windows/WSL 
**NPM version:** … 
6.14.8
**Node version:** …
14.9.0
**sgd version:** …
3.3.1
**git version:** …
2.17.1

### CODE REVIEW

File: lib/utils/repoGitDiff.js [hunk 1]
```diff
@@ -3,22 +3,31 @@ const childProcess = require('child_process')
...
 
 const fullDiffParams = ['diff', '--name-status', '--no-renames']
+const ig = ignore()
 
 module.exports = (config, metadata) => {
   const { stdout: diff } = childProcess.spawnSync(
     'git',
     [...fullDiffParams, config.from, config.to],
-    { cwd: config.repo, encoding: gc.UTF8_ENCODING }
+    { cwd: config.repo, encoding: gc.UTF8_ENCODING}
```
Comment (scolladon) [inline lib/utils/repoGitDiff.js:17]: Remove the unused space

File: lib/utils/repoGitDiff.js [hunk 2]
```diff
@@ -3,22 +3,31 @@ const childProcess = require('child_process')
 const cpUtils = require('./childProcessUtils')
 const gc = require('./gitConstants')
 const os = require('os')
-const path = require('path')
+const fs = require('fs')
+const path = require('path');
```
Comment (scolladon) [inline lib/utils/repoGitDiff.js:7]: Remove ; at the end of the line

File: lib/utils/repoGitDiff.js [hunk 3]
```diff
@@ -3,22 +3,31 @@ const childProcess = require('child_process')
...
+const ignore = require('ignore')
```
Comment (scolladon) [inline lib/utils/repoGitDiff.js:8]: Order the package import alphabetically : 
cpUtil
gc
fs
os
path
ignore

File: lib/utils/repoGitDiff.js [hunk 4]
```diff
@@ -3,22 +3,31 @@ const childProcess = require('child_process')
...
+    ig.add(fs.readFileSync(config.ignore).toString())
+  }
+
   return treatResult(cpUtils.treatDataFromSpawn(diff), metadata)
 }
 
 const treatResult = (repoDiffResult, metadata) =>
   repoDiffResult
     .split(os.EOL)
+    .filter(line => (line ? !ig.ignores(line) : false))
```
Comment (scolladon) [inline lib/utils/repoGitDiff.js:30]: `
.filter(line => !ig.ignores(line))
`
Could work ?

File: __tests__/unit/lib/utils/repoGitDiff.test.js [hunk 5]
```diff
@@ -3,7 +3,7 @@ const repoGitDiff = require('../../../../lib/utils/repoGitDiff')
 const child_process = require('child_process')
 jest.mock('child_process', () => ({ spawnSync: jest.fn() }))
 jest.mock('fs-extra')
-jest.mock('fs')
+//jest.mock('fs')
```
Comment (scolladon) [inline __tests__/unit/lib/utils/repoGitDiff.test.js:6]: Is it really working without this mock ?
If yes just remove the line plz

File: __tests__/unit/lib/utils/repoGitDiff.test.js [hunk 6]
```diff
@@ -12,25 +12,26 @@ describe(`test if repoGitDiff`, () => {
       stdout: '',
     }))
     // eslint-disable-next-line no-undef
-    const work = repoGitDiff({ output: '', repo: '' }, globalMetadata)
+    const work = repoGitDiff({ output: '', repo: '', ignore: '__tests__/unit/lib/utils/.forceignore' }, globalMetadata)
```
Comment (scolladon) [inline __tests__/unit/lib/utils/repoGitDiff.test.js:15]: use a constant at the beginning of the file for the path to the forceignore file plz

File: __tests__/unit/lib/utils/repoGitDiff.test.js [hunk 7]
```diff
@@ -12,25 +12,26 @@ describe(`test if repoGitDiff`, () => {
...
     const output = [
-      'D      force-app/main/default/objects/Account/fields/awesome.field-meta.xml',
+      'D      force-app/main/default/objects/Account/fields/awesome.field-meta.xml'
     ]
     child_process.spawnSync.mockImplementation(() => ({
       stdout: output[0],
     }))
     // eslint-disable-next-line no-undef
-    const work = repoGitDiff({ output: '', repo: '' }, globalMetadata)
+    const work = repoGitDiff({ output: '', repo: '', ignore: '__tests__/unit/lib/utils/.forceignore' }, globalMetadata)
```
Comment (scolladon) [inline __tests__/unit/lib/utils/repoGitDiff.test.js:28]: use a constant at the beginning of the file for the path to the forceignore file plz

File: __tests__/unit/lib/utils/repoGitDiff.test.js [hunk 8]
```diff
@@ -52,6 +53,20 @@ describe(`test if repoGitDiff`, () => {
...
 
+  test('can filter ignored files', () => {
+    const output = [
+      'M      force-app/main/default/lwc/jsconfig.json'
+    ]
+    child_process.spawnSync.mockImplementation(() => ({
+      stdout: output[0]
+    }))
+    // eslint-disable-next-line no-undef
+    const work = repoGitDiff({ output: '', repo: '', ignore: '__tests__/unit/lib/utils/.forceignore' }, globalMetadata)
```
Comment (scolladon) [inline __tests__/unit/lib/utils/repoGitDiff.test.js:64]: use a constant at the beginning of the file for the path to the forceignore file plz

File: __tests__/unit/lib/utils/repoGitDiff.test.js [hunk 9]
```diff
@@ -42,7 +43,7 @@ describe(`test if repoGitDiff`, () => {
 
   test('can resolve file copy when file is modified', () => {
     const output = [
-      'M      force-app/main/default/objects/Account/fields/awesome.field-meta.xml',
+      'M      force-app/main/default/objects/Account/fields/awesome.field-meta.xml'
```
Comment (scolladon) [inline __tests__/unit/lib/utils/repoGitDiff.test.js:46]: Leave the ',' at the end of the line
ecmascript support that

File: __tests__/unit/lib/utils/repoGitDiff.test.js [hunk 10]
```diff
@@ -12,25 +12,26 @@ describe(`test if repoGitDiff`, () => {
...
     // eslint-disable-next-line no-undef
-    const work = repoGitDiff({ output: '', repo: '' }, globalMetadata)
+    const work = repoGitDiff({ output: '', repo: '', ignore: '__tests__/unit/lib/utils/.forceignore' }, globalMetadata)
     expect(work).toMatchObject(output)
   })
 
   test('can resolve file copy when new file is added', () => {
     const output = [
-      'A      force-app/main/default/objects/Account/fields/awesome.field-meta.xml',
+      'A      force-app/main/default/objects/Account/fields/awesome.field-meta.xml'
```
Comment (scolladon) [inline __tests__/unit/lib/utils/repoGitDiff.test.js:34]: Leave the ',' at the end of the line
ecmascript support that

File: __tests__/unit/lib/utils/.forceignore [hunk 11]
```diff
@@ -0,0 +1,10 @@
+# List files or directories below to ignore them when running force:source:push, force:source:pull, and force:source:status
+# More information: https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_exclude_source.htm
+#
+
+package.xml
+**/jsconfig.json
+
+
+# LWC Jest
+**/__tests__/**
```
Comment (scolladon) [inline __tests__/unit/lib/utils/.forceignore:10]: I would prefer to move it under the __mocks__ folder 
(and adjust code reference to it in the repoGitDiff.test.js file)

**Review by scolladon — CHANGES_REQUESTED** [review]
Wonderfull suggestion, I really like it.

Please @jjulicher consider this review

And also could you add a test case for the ignore in the cli.test.js test file please

I am open to discussion :)

File: README.md [hunk 12]
```diff
@@ -45,6 +45,7 @@ To see the full list and description of the sgd options, run `sgd --help`
 -f, --from [sha] commit sha from where the diff is done [git rev-list —max-parents=0 HEAD]
 -o, --output [dir] source package specific output [./output] (default: "./output")
 -a, --api-version [version] salesforce API version [48] (default: "48")
+-i, --ignore specify the ignore file (default: ".forceignore")
 -r, --repo [dir] git repository location [./repo] (default: "./repo")
 -d, --generate-delta generate delta files in [./output] folder
 -h, --help output usage information
```
Comment (scolladon) [inline README.md:51]: Add also a reference to the ignore module in the Built With section plz

File: lib/utils/repoGitDiff.js (hunk 4 above)
Comment (jjulicher) [reply lib/utils/repoGitDiff.js:30]: Unfortunately line was sometimes null/undefined and ignores would throw an error.

File: __tests__/unit/lib/utils/repoGitDiff.test.js (hunk 5 above)
Comment (jjulicher) [reply __tests__/unit/lib/utils/repoGitDiff.test.js]: Yes it did! 

Ran all tests & passed - needed it to test ignore loading a file properly. Didn't seem to cause any other issues 

will remove.

**Review by scolladon — CHANGES_REQUESTED** [review]
Please don't remove .gitignore (just don't change it)
use npm run lint:fix to make sure everything is compliant with the coding style
And review the other comments please

**Comment by scolladon** [comment]
In order to resolve the conflicts please rebase the branch on top of origin/master

File: lib/utils/repoGitDiff.js (hunk 4 above)
Comment (scolladon) [reply lib/utils/repoGitDiff.js:31]: So we could rewrite it like this : 
`
.filter(line => line && !ig.ignores(line))
`
?

File: lib/utils/repoGitDiff.js [hunk 13]
```diff
@@ -1,24 +1,34 @@
...
-const childProcess = require('child_process')
 const cpUtils = require('./childProcessUtils')
+const childProcess = require('child_process')
 const gc = require('./gitConstants')
+const fs = require('fs')
+const ignore = require('ignore')
 const os = require('os')
 const path = require('path')
 
+
```
Comment (scolladon) [inline lib/utils/repoGitDiff.js:10]: Linting issue here

File: lib/utils/repoGitDiff.js (hunk 4 above)
Comment (jjulicher) [reply lib/utils/repoGitDiff.js:31]: ahh yes of course. that should work.  I'll update.

File: README.md (hunk 12 above)
Comment (jjulicher) [reply README.md:51]: Done.

File: lib/utils/repoGitDiff.js [hunk 14]
```diff
@@ -1,24 +1,34 @@
...
+    ig.add(fs.readFileSync(config.ignore).toString())
+  }
+
   return treatResult(cpUtils.treatDataFromSpawn(diff), metadata)
 }
 
 const treatResult = (repoDiffResult, metadata) =>
   repoDiffResult
     .split(os.EOL)
+    .filter(line => line && !ig.ignores(line))
```
Comment (scolladon) [inline lib/utils/repoGitDiff.js:31]: in fact we should use [double negation](https://riptutorial.com/javascript/example/3047/double-negation----x-) check here for the code to be more robust.

Could you change it to 
`
.filter(line => !!line && !ig.ignores(line))
`

**Comment by scolladon** [comment]
We may have an issue here...
It seems commit from other branch have been added here...

**Comment by jjulicher** [comment]
Oh crap. Sorry it's late after a long week.... origin/master is from my
repo which was already merged with all the changes for my CI pipeline.

I'm going to have to create a new branch aren't I... 🙄😪 And reintegrate
all my changes.....

**Comment by scolladon** [comment]
No worries, 

I am rebasing the work right now and I am applying the fixes required.
I'll let you have a look before merging.

Cheers

**Review by scolladon — DISMISSED** [review]

**Comment by scolladon** [comment]
@jjulicher everything looks ok for me, is it ok for you too ?

@mehdisfdc are you ok to merge this feature ?

**Comment by jjulicher** [comment]
> No worries,
> 
> I am rebasing the work right now and I am applying the fixes required.
> I'll let you have a look before merging.
> 
> Cheers

Oh thanks!

**Comment by jjulicher** [comment]
> @jjulicher everything looks ok for me, is it ok for you too ?
> 
> @mehdisfdc are you ok to merge this feature ?

Looks good to me.

**Review by scolladon — APPROVED** [review]

**Comment by mehdisfdc** [comment]
> @jjulicher everything looks ok for me, is it ok for you too ?
> 
> @mehdisfdc are you ok to merge this feature ?

👍 

Thank you @jjulicher for your contribution!

**Pull request merged**